subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2019,2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/zephyr.h>
+#include <soc.h>
+#include <adsp_memory.h>
+
+/*
+ * A lock is needed as log_process() and log_panic() have no internal locks
+ * to prevent concurrency. Meaning if log_process is called after
+ * log_panic was called previously, log_process may happen from another
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock mtrace_lock;
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_ADSP_MTRACE_OUTPUT_DEFAULT;
+
+static adsp_mtrace_log_hook_t mtrace_hook;
+
+static bool mtrace_active;
+
+/*
+ * SRAM window for mtrace is organized in 16 (equal sized) pages:
+ *
+ *	------------------------
+ *	| Page0  - descriptors |
```

Comment:
Seems like 4k of memory for a handful of descriptors is a little wasteful?

Reply:
@andyross Indeed. But it is part of the protocol (as also implemented here https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-mtrace.c ) . I believe there is rationale to keep the descriptors on a separate page to enable separate R/W access control to host. The host can only read the descriptors, while the data slots are also writable by host. It's still a bit klunky, but it is what it is. We can always just use the ADSP backend if we don't need to comply with the protocol and use more of the SRAM.